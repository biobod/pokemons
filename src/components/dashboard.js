import React, { Component } from 'react';
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import PokemonCard from './pokemonCard'
import { fetchPokemon, fetchPokemonsCount } from './../redux/actions'
import ButtonsSection from './buttonsSection'

class Dashboard extends Component {
  componentWillMount() {
    const { id, getPokemon, getPokemonsCount } = this.props
    getPokemonsCount()
    getPokemon(id)
  }
  componentDidMount() {
    window.addEventListener('wheel', this.listenScroll)
  }
  componentWillUnmount() {
    window.removeEventListener('wheel', this.listenScroll);
  }
  
  showNextPokemon = () => {
    const { id, getPokemon, pokemonsCount } = this.props
    let nextId = +id + 1
    if (nextId > pokemonsCount) {
      nextId = 1
    }
    getPokemon(nextId)
  }
  showPreviousPokemon = () => {
    const { id, getPokemon, pokemonsCount } = this.props
    let previousId = id - 1
    if (previousId <= 0) {
      previousId = pokemonsCount
    }
    getPokemon(previousId)
  }
  listenScroll = (e) => {
    const { isLoading } = this.props
    if (isLoading) {
      return
    }
    if (e.deltaY > 0) {
      this.showPreviousPokemon()
    } else {
      this.showNextPokemon()
    }
  }
  
  render() {
    const {
      showPreviousPokemon, showNextPokemon,
      props: { pokemon, isError, isLoading }
    } = this
    if (isError) {
      return <div>Sorry, something go wrong</div>
    }
    return (
      <div>
        <ButtonsSection
          showNextPokemon={showNextPokemon}
          showPreviousPokemon={showPreviousPokemon}
          isLoading={isLoading}
        >
          {pokemon && <PokemonCard pokemon={pokemon} isLoading={isLoading} />}
        </ButtonsSection>
      </div>
    )
  }
}

Dashboard.propTypes = {
  pokemon: PropTypes.shape(),
  isError: PropTypes.bool,
  getPokemon: PropTypes.func.isRequired,
  getPokemonsCount: PropTypes.func.isRequired,
  id: PropTypes.number.isRequired,
  isLoading: PropTypes.bool.isRequired,
  pokemonsCount: PropTypes.number.isRequired
};

export default connect(state => ({
  id: state.currentPokemonId,
  pokemon: state.pokemon,
  isError: state.error,
  isLoading: state.isLoading,
  pokemonsCount: state.pokemonsCount,
}), { getPokemon: fetchPokemon, getPokemonsCount: fetchPokemonsCount })(Dashboard);
