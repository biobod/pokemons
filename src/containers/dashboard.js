import React, { Component } from 'react';
import PropTypes from 'prop-types'

import { connect } from 'react-redux'
import PokemonCard from './pokemonCard'
import { fetchPokemon } from './../redux/actions'
import ArrowWrapper from './cardArrowWrapper'

class Dashboard extends Component {
  componentWillMount = () => {
    const { id, getPokemon } = this.props
    getPokemon(id)
  }
  
  showNextPokemon = () => {
    const { id, getPokemon } = this.props
    getPokemon(+id + 1)
  }
  showPreviousPokemon = () => {
    const { id, getPokemon } = this.props
    getPokemon(id - 1)
  }
  
  render() {
    const {
      showPreviousPokemon, showNextPokemon,
      props: { pokemon, isError }
    } = this
    if (isError) {
      return <div>Sorry, something go wrong</div>
    }
    return (
      <div>
        <ArrowWrapper showNextPokemon={showNextPokemon} showPreviousPokemon={showPreviousPokemon}>
          {pokemon ? <PokemonCard
          name={pokemon.name}
          image={pokemon.sprites.front_default}
          />
        : <span>loading</span>
      }
        </ArrowWrapper>
      </div>
    )
  }
}

export default connect(state => ({
  id: state.currentPokemonId,
  pokemon: state.pokemon,
  isError: state.error,
}), { getPokemon: fetchPokemon })(Dashboard);
