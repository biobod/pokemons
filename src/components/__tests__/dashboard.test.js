import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

import Adapter from 'enzyme-adapter-react-16';

import { Dashboard } from '../dashboard';

Enzyme.configure({ adapter: new Adapter() });


describe('Dashboard', () => {
  test('should render correctly', () => {
    const component = renderer.create(<Dashboard
        getPokemon={() => {}}
        getPokemonsCount={() => {}}
        id={123}
        isLoading={false}
        pokemonsCount={700}
    />);
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot();
  });
});
