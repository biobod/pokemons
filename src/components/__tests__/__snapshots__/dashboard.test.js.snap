// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard should render correctly 1`] = `
<div>
  <div
    className="ButtonsSection-dashboardWrapper-1"
  >
    <div
      className="ButtonsSection-buttonSection-2"
    >
      <button
        className="MuiButtonBase-root-19 MuiButton-root-3 MuiButton-raised-9"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex="0"
        type="button"
      >
        <span
          className="MuiButton-label-4"
        >
          Previous
        </span>
        <span
          className="MuiTouchRipple-root-22"
        />
      </button>
      <button
        className="MuiButtonBase-root-19 MuiButton-root-3 MuiButton-raised-9"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex="0"
        type="button"
      >
        <span
          className="MuiButton-label-4"
        >
          Next
        </span>
        <span
          className="MuiTouchRipple-root-22"
        />
      </button>
    </div>
  </div>
</div>
`;
